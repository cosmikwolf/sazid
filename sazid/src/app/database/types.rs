use diesel::sql_types::*;
use serde::{Deserialize, Serialize};
use std::fmt;

use super::schema::*;
use crate::app::{errors::SazidError, session_config::SessionConfig};
use async_openai::types::ChatCompletionRequestMessage;
use diesel::{expression::ValidGrouping, prelude::*};
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use diesel_json;
use pgvector::Vector;

#[derive(
  Serialize,
  Deserialize,
  Queryable,
  Selectable,
  Debug,
  Clone,
  Identifiable,
  PartialEq,
  ValidGrouping,
)]
#[diesel(table_name = sessions)]
pub struct QueryableSession {
  pub id: i64,
  pub config: diesel_json::Json<SessionConfig>,
  pub summary: Option<String>,
}

#[derive(
  Serialize,
  Queryable,
  Selectable,
  Debug,
  Clone,
  Identifiable,
  PartialEq,
  ValidGrouping,
  Associations,
)]
#[diesel(table_name = messages)]
#[diesel(belongs_to(QueryableSession, foreign_key = session_id))]
pub struct QueryableMessage {
  id: i64,
  data: diesel_json::Json<ChatCompletionRequestMessage>,
  #[serde(skip)]
  embedding: Vector,
  session_id: i64,
}

#[derive(Insertable, Debug, Clone, PartialEq, AsChangeset)]
#[diesel(table_name = sessions)]
pub struct InsertableSession {
  pub config: diesel_json::Json<SessionConfig>,
  pub summary: Option<String>,
}

#[derive(
  Serialize,
  Queryable,
  Selectable,
  Debug,
  Clone,
  Identifiable,
  PartialEq,
  Associations,
  ValidGrouping,
)]
#[diesel(belongs_to(FileEmbedding))]
#[diesel(table_name = embedding_pages)]
pub struct EmbeddingPage {
  id: i64,
  content: String,
  checksum: String,
  page_number: i32,
  #[serde(skip)]
  pub embedding: Vector,
  file_embedding_id: i64,
}

#[derive(Insertable, Debug, Clone, PartialEq, AsChangeset)]
#[diesel(table_name = embedding_pages)]
pub struct InsertablePage {
  pub content: String,
  pub page_number: i32,
  pub checksum: String,
  pub embedding: Vector,
}

#[derive(
  Serialize,
  Queryable,
  Selectable,
  Debug,
  Clone,
  PartialEq,
  Identifiable,
  AsChangeset,
  ValidGrouping,
)]
#[diesel(table_name = file_embeddings)]
pub struct FileEmbedding {
  id: i64,
  pub filepath: String,
  checksum: String,
}

#[derive(Insertable, Debug, Clone, PartialEq, AsChangeset)]
#[diesel(table_name = file_embeddings)]
pub struct InsertableFileEmbedding {
  pub filepath: String,
  pub checksum: String,
}

#[derive(Queryable, Selectable, Debug, Clone, PartialEq, Identifiable, AsChangeset)]
#[diesel(table_name = tags)]
pub struct Tag {
  id: i64,
  tag: String,
}

#[derive(Serialize)]
pub struct FileWithPages {
  #[serde(flatten)]
  pub file: FileEmbedding,
  pub pages: Vec<EmbeddingPage>,
}

#[derive(Queryable, Selectable, Debug, Clone, PartialEq, AsChangeset)]
#[diesel(table_name = tags)]
pub struct InsertableTag {
  tag: String,
}

#[derive(Debug, Clone, PartialEq, Queryable, Selectable, Associations, Identifiable)]
#[diesel(belongs_to(FileEmbedding))]
#[diesel(table_name = embedding_tags)]
#[diesel(primary_key(file_embedding_id, tag_id))]
pub struct EmbeddingTag {
  file_embedding_id: i64,
  tag_id: i64,
}

impl FileEmbedding {
  pub async fn page_count(&self, conn: &mut AsyncPgConnection) -> Result<usize, SazidError> {
    let all_pages = EmbeddingPage::belonging_to(self)
      .select(EmbeddingPage::as_select())
      .load::<EmbeddingPage>(conn)
      .await?;
    Ok(all_pages.len())
  }
}

impl EmbeddingPage {
  pub async fn get_embedding_from_page(
    &self,
    conn: &mut AsyncPgConnection,
  ) -> Result<FileEmbedding, SazidError> {
    let embedding = file_embeddings::table
      .filter(file_embeddings::id.eq(self.file_embedding_id))
      .select(FileEmbedding::as_select())
      .first::<FileEmbedding>(conn)
      .await?;
    Ok(embedding)
  }
  pub async fn get_next_page(
    &self,
    conn: &mut AsyncPgConnection,
  ) -> Result<Option<EmbeddingPage>, SazidError> {
    let next_page = embedding_pages::table
      .filter(embedding_pages::file_embedding_id.eq(self.file_embedding_id))
      .filter(embedding_pages::page_number.gt(self.page_number))
      .order_by(embedding_pages::page_number.asc())
      .select(EmbeddingPage::as_select())
      .first::<EmbeddingPage>(conn)
      .await
      .optional()?;
    Ok(next_page)
  }

  pub async fn get_previous_page(
    &self,
    conn: &mut AsyncPgConnection,
  ) -> Result<Option<EmbeddingPage>, SazidError> {
    let previous_page = embedding_pages::table
      .filter(embedding_pages::file_embedding_id.eq(self.file_embedding_id))
      .filter(embedding_pages::page_number.lt(self.page_number))
      .order_by(embedding_pages::page_number.desc())
      .select(EmbeddingPage::as_select())
      .first::<EmbeddingPage>(conn)
      .await
      .optional()?;
    Ok(previous_page)
  }
}

#[derive(QueryableByName, Debug)]
pub struct PgVectorIndexInfo {
  #[diesel(sql_type = Int4)]
  pub indexrelid: i32,
  #[diesel(sql_type = Text)]
  pub indexname: String,
  #[diesel(sql_type = Bool)]
  pub indexing: bool,
  #[diesel(sql_type = Int4)]
  pub idx_tuples: i32,
  #[diesel(sql_type = Int4)]
  pub idx_sealed_len: i32,
  #[diesel(sql_type = Int4)]
  pub idx_growing_len: i32,
  #[diesel(sql_type = Int4)]
  pub idx_write: i32,
  #[diesel(sql_type = Text)]
  pub idx_config: String,
}

// impl fmt::Display for FileWithPages {
//   fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
//     write!(f, "{}", self)
//   }
// }
impl fmt::Display for EmbeddingPage {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    write!(
      f,
      "EmbeddingPage(page_number: {}\t {} bytes,  first line: {}",
      self.page_number,
      self.content.as_bytes().len(),
      self.content.lines().next().unwrap()
    )
  }
}

impl fmt::Display for FileEmbedding {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    write!(f, "TextFileEmbedding(filename: {}\tchecksum: {}) ", self.filepath, self.checksum,)
  }
}

impl Iterator for FileEmbedding {
  type Item = FileEmbedding;
  fn next(&mut self) -> Option<Self::Item> {
    Some(self.clone())
  }
}
