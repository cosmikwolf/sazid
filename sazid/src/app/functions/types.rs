use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct FunctionProperties {
  #[serde(skip)]
  pub name: String,
  #[serde(skip)]
  pub required: bool,
  #[serde(rename = "type")]
  pub property_type: String,
  pub description: Option<String>,
  #[serde(rename = "enum", default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub enum_values: Option<Vec<String>>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct FunctionParameters {
  #[serde(rename = "type")]
  pub param_type: String,
  pub required: Vec<String>,
  pub properties: std::collections::HashMap<String, FunctionProperties>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct FunctionCall {
  pub name: String,
  pub description: Option<String>,
  pub parameters: Option<FunctionParameters>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Commands {
  pub commands: Vec<FunctionCall>,
}
