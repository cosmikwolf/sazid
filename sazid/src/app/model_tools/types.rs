use serde::{Deserialize, Serialize, Serializer};
use serde_json::Value;
use std::collections::HashMap;
use std::path::PathBuf;

fn serialize_parameters<S>(
  properties: &std::collections::HashMap<String, FunctionProperty>,
  serializer: S,
) -> Result<S::Ok, S::Error>
where
  S: Serializer,
{
  use serde::ser::SerializeStruct;
  let mut state = serializer.serialize_struct("Properties", 3)?;
  state.serialize_field("type", "object")?;
  state.serialize_field("properties", properties)?;
  let required: Vec<String> = properties
    .iter()
    .filter_map(|(key, value)| match value {
      FunctionProperty::Bool { required, .. }
      | FunctionProperty::Number { required, .. }
      | FunctionProperty::String { required, .. }
      | FunctionProperty::Pattern { required, .. }
      | FunctionProperty::Null { required, .. }
      | FunctionProperty::PathBuf { required, .. }
      | FunctionProperty::Array { required, .. }
      | FunctionProperty::Integer { required, .. } => {
        if *required {
          Some(key.clone())
        } else {
          None
        }
      },
      FunctionProperty::Parameters { .. } => None,
    })
    .collect();
  state.serialize_field("required", &required)?;
  state.end()
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "type")]
pub enum FunctionProperty {
  #[serde(rename = "object")]
  Parameters {
    #[serde(flatten, serialize_with = "serialize_parameters")]
    properties: std::collections::HashMap<String, FunctionProperty>,
  },
  #[serde(rename = "boolean")]
  Bool {
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip)]
    required: bool,
  },
  #[serde(rename = "number")]
  Number {
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip)]
    required: bool,
  },
  #[serde(rename = "string")]
  String {
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip)]
    required: bool,
  },
  #[serde(rename = "string")]
  Pattern {
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip)]
    required: bool,
  },
  #[serde(rename = "null")]
  Null {
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip)]
    required: bool,
  },
  #[serde(rename = "string")]
  PathBuf {
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip)]
    required: bool,
  },
  #[serde(rename = "array")]
  Array {
    #[serde(rename = "items")]
    items: Box<FunctionProperty>,
    #[serde(rename = "minItems")]
    min_items: Option<usize>,
    #[serde(rename = "maxItems")]
    max_items: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip)]
    required: bool,
  },
  #[serde(rename = "number")]
  Integer {
    #[serde(rename = "minimum")]
    minimum: Option<i64>,
    #[serde(rename = "maximum")]
    maximum: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip)]
    required: bool,
  },
}

pub fn validate_arguments(
  arguments: HashMap<String, Value>,
  parameters: &FunctionProperty,
  workspace_root: Option<&PathBuf>,
) -> Result<HashMap<String, Value>, String> {
  let properties = if let FunctionProperty::Parameters { properties } = parameters {
    properties
  } else {
    return Err("parameters must be FunctionProperty::Parameters".to_string());
  };

  let mut validated_args: HashMap<String, Value> = HashMap::new();

  for (name, property) in properties {
    let arg_value = arguments.get(name);

    match (arg_value, property) {
      (Some(value), FunctionProperty::Bool { required: _, .. }) => {
        if let Some(v) = value.as_bool() {
          validated_args.insert(name.clone(), Value::Bool(v));
        } else {
          return Err(format!(
            "Invalid type for argument '{}'. Expected: bool, Found: {:?}",
            name, value
          ));
        }
      },
      (Some(value), FunctionProperty::Number { required: _, .. }) => {
        if let Some(v) = value.as_f64() {
          validated_args.insert(name.clone(), serde_json::Number::from_f64(v).into());
        } else {
          return Err(format!(
            "Invalid type for argument '{}'. Expected: number, Found: {:?}",
            name, value
          ));
        }
      },
      (Some(value), FunctionProperty::String { required: _, .. }) => {
        if let Some(v) = value.as_str() {
          validated_args.insert(name.clone(), Value::String(v.to_string()));
        } else {
          return Err(format!(
            "Invalid type for argument '{}'. Expected: string, Found: {:?}",
            name, value
          ));
        }
      },
      (Some(value), FunctionProperty::Array { items, min_items, max_items, required: _, .. }) => {
        if let Some(arr) = value.as_array() {
          if let Some(min) = min_items {
            if arr.len() < *min {
              return Err(format!(
                "Array length for argument '{}' is below the minimum. Expected: {}, Found: {}",
                name,
                min,
                arr.len()
              ));
            }
          }
          if let Some(max) = max_items {
            if arr.len() > *max {
              return Err(format!(
                "Array length for argument '{}' exceeds the maximum. Expected: {}, Found: {}",
                name,
                max,
                arr.len()
              ));
            }
          }
          let mut validated_arr = Vec::new();
          for item in arr {
            if is_valid_type(item, items) {
              validated_arr.push(item.clone());
            } else {
              return Err(format!(
                "Invalid type for array item in argument '{}'. Expected: {:?}, Found: {:?}",
                name, items, item
              ));
            }
          }
          validated_args.insert(name.clone(), Value::Array(validated_arr));
        } else {
          return Err(format!(
            "Invalid type for argument '{}'. Expected: array, Found: {:?}",
            name, value
          ));
        }
      },
      (Some(value), FunctionProperty::Null { required: _, .. }) => {
        if value.is_null() {
          validated_args.insert(name.clone(), Value::Null);
        } else {
          return Err(format!(
            "Invalid type for argument '{}'. Expected: null, Found: {:?}",
            name, value
          ));
        }
      },
      (Some(value), FunctionProperty::Integer { minimum, maximum, required: _, .. }) => {
        if let Some(v) = value.as_i64() {
          if let Some(min) = minimum {
            if v < *min {
              return Err(format!(
                "Value for argument '{}' is below the minimum. Expected: {}, Found: {}",
                name, min, v
              ));
            }
          }
          if let Some(max) = maximum {
            if v > *max {
              return Err(format!(
                "Value for argument '{}' exceeds the maximum. Expected: {}, Found: {}",
                name, max, v
              ));
            }
          }
          validated_args.insert(name.clone(), Value::Number(v.into()));
        } else {
          return Err(format!(
            "Invalid type for argument '{}'. Expected: integer, Found: {:?}",
            name, value
          ));
        }
      },
      (Some(value), FunctionProperty::Pattern { required: _, .. }) => {
        if let Some(pattern_str) = value.as_str() {
          match regex::Regex::new(pattern_str) {
            Ok(_regex) => {
              validated_args.insert(name.clone(), Value::String(pattern_str.to_string()));
            },
            Err(err) => {
              return Err(format!(
                "Invalid regular expression pattern for argument '{}'. Error: {}",
                name, err
              ));
            },
          }
        } else {
          return Err(format!(
            "Invalid type for argument '{}'. Expected: string (regex pattern), Found: {:?}",
            name, value
          ));
        }
      },
      (Some(value), FunctionProperty::PathBuf { required: _, .. }) => {
        if workspace_root.is_none() {
          return Err(format!("Workspace root is required to validate path argument '{}'", name));
        }

        if let Some(path_str) = value.as_str() {
          let path = PathBuf::from(path_str);

          // return an error if path is not within workspace
          match workspace_root {
            Some(workspace_dir) => {
              if !path.starts_with(workspace_dir) {
                return Err("cannot read files outside of the current working directory".into());
              }
            },
            None => return Err("cannot create files without a workspace set".into()),
          }

          if !path.is_absolute() {
            let absolute_path = workspace_root.unwrap().join(path);
            if absolute_path.exists() {
              validated_args
                .insert(name.clone(), Value::String(absolute_path.to_string_lossy().into_owned()));
            } else {
              return Err(format!(
                "Invalid path for argument '{}'. Path does not exist: {:?}",
                name, absolute_path
              ));
            }
          } else if path.exists() {
            validated_args.insert(name.clone(), Value::String(path.to_string_lossy().into_owned()));
          } else {
            return Err(format!(
              "Invalid path for argument '{}'. Path does not exist: {:?}",
              name, path
            ));
          }
        } else {
          return Err(format!(
            "Invalid type for argument '{}'. Expected: string (path), Found: {:?}",
            name, value
          ));
        }
      },
      (None, FunctionProperty::Bool { required: true, .. })
      | (None, FunctionProperty::Number { required: true, .. })
      | (None, FunctionProperty::String { required: true, .. })
      | (None, FunctionProperty::Array { required: true, .. })
      | (None, FunctionProperty::Null { required: true, .. })
      | (None, FunctionProperty::Integer { required: true, .. })
      | (None, FunctionProperty::Pattern { required: true, .. })
      | (None, FunctionProperty::PathBuf { required: true, .. }) => {
        return Err(format!("Missing required argument: '{}'", name));
      },
      (None, FunctionProperty::Bool { required: false, .. })
      | (None, FunctionProperty::Number { required: false, .. })
      | (None, FunctionProperty::String { required: false, .. })
      | (None, FunctionProperty::Array { required: false, .. })
      | (None, FunctionProperty::Null { required: false, .. })
      | (None, FunctionProperty::Integer { required: false, .. })
      | (None, FunctionProperty::Pattern { required: false, .. })
      | (None, FunctionProperty::PathBuf { required: false, .. }) => {
        // Skip optional arguments that are not provided
      },
      (_, FunctionProperty::Parameters { .. }) => {
        // Skip validation for nested properties
      },
    }
  }

  Ok(validated_args)
}

pub fn get_validated_argument<T: serde::de::DeserializeOwned>(
  validated_arguments: &HashMap<String, Value>,
  key: &str,
) -> Option<T> {
  validated_arguments.get(key).and_then(|value| serde_json::from_value(value.clone()).ok())
}

fn is_valid_type(value: &Value, expected_type: &FunctionProperty) -> bool {
  matches!(
    (value, expected_type),
    (Value::Bool(_), FunctionProperty::Bool { .. })
      | (Value::Number(_), FunctionProperty::Number { .. })
      | (Value::String(_), FunctionProperty::String { .. })
      | (Value::Array(_), FunctionProperty::Array { .. })
      | (Value::Null, FunctionProperty::Null { .. })
      | (Value::Number(_), FunctionProperty::Integer { .. })
      | (Value::String(_), FunctionProperty::Pattern { .. })
      | (Value::String(_), FunctionProperty::PathBuf { .. })
  )
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ToolCall {
  pub name: String,
  pub description: Option<String>,
  pub parameters: Option<FunctionProperty>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Commands {
  pub commands: Vec<ToolCall>,
}
