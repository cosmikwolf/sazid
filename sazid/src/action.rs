use std::path::PathBuf;

use crate::{
  app::{
    database::types::QueryableSession,
    lsi::query::LsiQuery,
    messages::ChatMessage,
    session_config::{SessionConfig, WorkspaceParams},
  },
  components::data_manager::DataManagerAction,
};
use async_openai::types::{
  ChatCompletionMessageToolCall, ChatCompletionRequestMessage, ChatCompletionTool,
};
use helix_lsp::Call;
use lsp_types::{DocumentSymbol, TextDocumentIdentifier};
use serde::{Deserialize, Serialize, Serializer};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum ToolType {
  LsiQuery(LsiQuery),
  Generic(i64, String),
}
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum SessionAction {
  LspCheckServerNotifications,
  LspServerMessageReceived((usize, Call)),
  LspSymbolQuery(LsiQuery),
  CreateSession(SessionConfig),
  LoadSession(i64),
  SetTestToolResponse(ToolType, String),
  ToolCallComplete(ToolType, String),
  ToolCallError(ToolType, String),

  CreateLoadSessionResponse(QueryableSession),
  AddMessageEmbedding(i64, i64, ChatCompletionRequestMessage),
  MessageEmbeddingSuccess(i64),
  RequestRelatedMessages(i64, String, bool),
  SubmitInput(String),
  ExecuteCommand(String),
  CommandResult(String),
  RequestChatCompletion(),
  AddMessage(i64, ChatMessage),
  UpdateMessage(ChatCompletionRequestMessage, i64),
  ReloadMessages(Vec<(i64, ChatCompletionRequestMessage)>),
  UpdateStatus(Option<String>),
  UpdateToolList(i64, Vec<ChatCompletionTool>),

  SaveSession,

  LsiAction(LsiAction),
  DataManagerAction(DataManagerAction),
  ChatToolAction(ChatToolAction),

  Error(String),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum LsiAction {
  #[serde(serialize_with = "serialize_boxed_session_action")]
  SessionAction(Box<SessionAction>),
  ChatToolResponse(Box<ChatToolAction>),
  AddWorkspace(WorkspaceParams),
  QueryWorkspaceSymbols(LsiQuery),
  GetWorkspaceFiles(LsiQuery),
  ReplaceSymbolText(String, LsiQuery),
  ReadSymbolSource(LsiQuery),
  GoToSymbolDefinition(LsiQuery),
  GoToSymbolDeclaration(LsiQuery),
  GoToTypeDefinition(LsiQuery),
  GetDiagnostics(LsiQuery),
  UpdateWorkspaceFileSymbols(PathBuf, TextDocumentIdentifier, Vec<DocumentSymbol>),
  RequestWorkspaceFileSymbols(PathBuf, TextDocumentIdentifier, usize),
  Error(String),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum ChatToolAction {
  UpdateConfig(i64, Box<SessionConfig>),
  CallTool(ChatCompletionMessageToolCall, i64),
  CompleteToolCall(String, ChatCompletionMessageToolCall, i64),
  #[serde(serialize_with = "serialize_boxed_session_action")]
  SessionAction(Box<SessionAction>),
  LsiRequest(Box<LsiAction>),
  LsiQueryResponse(String, String),
  ToolListRequest(i64),
  ToolListResponse(i64, Vec<ChatCompletionTool>),
  Error(String),
}

pub fn serialize_boxed_session_action<S>(
  action: &SessionAction,
  serializer: S,
) -> Result<S::Ok, S::Error>
where
  S: Serializer,
{
  action.serialize(serializer)
}
//
// impl<'de> Deserialize<'de> for Action {
//   fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
//   where
//     D: Deserializer<'de>,
//   {
//     struct ActionVisitor;
//
//     impl<'de> Visitor<'de> for ActionVisitor {
//       type Value = Action;
//
//       fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
//         formatter.write_str("a valid string representation of Action")
//       }
//
//       fn visit_str<E>(self, value: &str) -> Result<Action, E>
//       where
//         E: de::Error,
//       {
//         match value {
//           "Tick" => Ok(Action::Tick),
//           "Render" => Ok(Action::Render),
//           "Suspend" => Ok(Action::Suspend),
//           "Resume" => Ok(Action::Resume),
//           "Quit" => Ok(Action::Quit),
//           "Refresh" => Ok(Action::Refresh),
//           "Help" => Ok(Action::Help),
//           "EnterInsert" => Ok(Action::EnterInsert),
//           "EnterNormal" => Ok(Action::EnterNormal),
//           data if data.starts_with("Error(") => {
//             let error_msg =
//               data.trim_start_matches("Error(").trim_end_matches(')');
//             Ok(Action::Error(error_msg.to_string()))
//           },
//           data if data.starts_with("Resize(") => {
//             let parts: Vec<&str> = data
//               .trim_start_matches("Resize(")
//               .trim_end_matches(')')
//               .split(',')
//               .collect();
//             if parts.len() == 2 {
//               let width: u16 = parts[0].trim().parse().map_err(E::custom)?;
//               let height: u16 = parts[1].trim().parse().map_err(E::custom)?;
//               Ok(Action::Resize(width, height))
//             } else {
//               Err(E::custom(format!("Invalid Resize format: {}", value)))
//             }
//           },
//           _ => Err(E::custom(format!("Unknown Action variant: {}", value))),
//         }
//       }
//     }
//
//     deserializer.deserialize_str(ActionVisitor)
//   }
// }
