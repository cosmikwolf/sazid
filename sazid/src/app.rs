use serde::{Deserialize, Serialize};

pub mod color_math;
pub mod consts;
pub mod database;
pub mod errors;
pub mod gpt_interface;
pub mod helpers;
pub mod lsi;
pub mod markdown;
pub mod messages;
pub mod model_tools;
pub mod request_validation;
pub mod session_config;
pub mod tools;
pub mod treesitter;
pub mod types;

#[derive(Default, Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Mode {
  #[default]
  Home,
}
