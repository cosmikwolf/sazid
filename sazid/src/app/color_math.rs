use ratatui::style::Color;

// Compute both rainbow color and its hue-inverted counterpart.
pub fn get_rainbow_and_inverse_colors(
  step: u32,
  total_steps: u32,
) -> (Color, Color) {
  let hue = step as f32 / total_steps as f32;
  let inverse_hue = (hue + 0.5) % 1.0; // Add 0.5 to invert the hue in the spectrum and use modulo to wrap around

  let rainbow_color = hsv_to_rgb(hue, 1.0, 1.0);
  let inverse_color = hsv_to_rgb(inverse_hue, 1.0, 1.0);

  // Map the floating point RGB values to 8-bit integers
  let rainbow_rgb = (
    (rainbow_color.0 * 255.0).round() as u8,
    (rainbow_color.1 * 255.0).round() as u8,
    (rainbow_color.2 * 255.0).round() as u8,
  );

  let inverse_rgb = (
    (inverse_color.0 * 255.0).round() as u8,
    (inverse_color.1 * 255.0).round() as u8,
    (inverse_color.2 * 255.0).round() as u8,
  );

  (
    Color::Rgb(rainbow_rgb.0, rainbow_rgb.1, rainbow_rgb.2),
    Color::Rgb(inverse_rgb.0, inverse_rgb.1, inverse_rgb.2),
  )
}

fn hsv_to_rgb(h: f32, s: f32, v: f32) -> (f32, f32, f32) {
  let i = (h * 6.0).floor() as i32;
  let f = h * 6.0 - i as f32;
  let p = v * (1.0 - s);
  let q = v * (1.0 - f * s);
  let t = v * (1.0 - (1.0 - f) * s);

  match i % 6 {
    0 => (v, t, p),
    1 => (q, v, p),
    2 => (p, v, t),
    3 => (p, q, v),
    4 => (t, p, v),
    _ => (v, p, q),
  }
}
