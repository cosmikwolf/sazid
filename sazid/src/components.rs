use core::result::Result;
// This is a sample comment
use helix_view::graphics::Rect;
use tokio::sync::mpsc::UnboundedSender;

use crate::{action::SessionAction, app::errors::SazidError, config::Config};

use tui::buffer::Buffer;

pub mod data_manager;
pub mod session;

pub trait Component {
  #[allow(unused_variables)]
  fn register_action_handler(
    &mut self,
    tx: UnboundedSender<SessionAction>,
  ) -> Result<(), SazidError> {
    Ok(())
  }
  #[allow(unused_variables)]
  fn register_config_handler(
    &mut self,
    config: Config,
  ) -> Result<(), SazidError> {
    Ok(())
  }
  fn init(&mut self, _area: Rect) -> Result<(), SazidError> {
    Ok(())
  }
  #[allow(unused_variables)]
  fn update(
    &mut self,
    action: SessionAction,
  ) -> Result<Option<SessionAction>, SazidError> {
    Ok(None)
  }
  fn draw(&mut self, b: &mut Buffer) -> Result<(), SazidError>;
}
