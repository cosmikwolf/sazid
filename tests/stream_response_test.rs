extern crate lazy_static;

#[cfg(test)]
mod tests {
  use async_openai::types::Role;
  use ntest::timeout;
  use sazid::action::Action;
  use sazid::app::types::{ChatMessage, ChatResponse, ChatResponseSingleMessage};
  use sazid::components::session::*;
  use tokio::sync::mpsc;

  #[tokio::test]
  #[timeout(10000)]
  pub async fn test_request_response() {
    let mut enter_processing_action_run = false;
    let mut process_response_action_run = false;
    let (tx, mut rx) = mpsc::unbounded_channel::<Action>();
    let mut session = Session::new();
    session.request_response("Hello World".to_string(), tx.clone());
    'outer: loop {
      while let Some(res) = rx.recv().await {
        match res {
          Action::EnterProcessing => {
            enter_processing_action_run = true;
          },
          Action::ProcessResponse(response) => {
            process_response_action_run = true;
            session.response_handler(tx.clone(), response.clone());
            if let ChatResponse::StreamResponse(message) = response {
              insta::assert_yaml_snapshot!(&message, { ".id" => "[id]", ".created"  => "[created]" });
            } else {
              panic!("Expected StreamResponse");
            };
          },
          Action::ExitProcessing => {
            // break;
            if let Some(ChatMessage::ChatCompletionResponseMessage(ChatResponseSingleMessage::StreamResponse(
              combined,
            ))) = session.data.messages.last()
            {
              assert!(process_response_action_run);
              assert!(enter_processing_action_run);
              insta::assert_yaml_snapshot!(&combined, { ".id" => "[id]", ".created"  => "[created]" });
              insta::assert_yaml_snapshot!(&session.data.messages.last().unwrap(), { ".id" => "[id]", ".created"  => "[created]" });
            } else {
              panic!(
                "Expected last transaction message to be StreamResponse {:#?}",
                session.data.messages.last_mut().unwrap()
              );
            }
            break 'outer;
          },
          Action::Update => {},
          Action::Render => {},
          _ => {
            panic!("Unexpected action {:#?}", res);
          },
        }
      }
    }
  }

  #[test]
  fn test_construct_chat_completion_request_message() {
    let mut session = Session::new();
    if let Ok(create_chat_completion_request_message_result) = session.add_chunked_chat_completion_request_messages(
      "testing testing 1 2 3",
      "sazid testing",
      Role::User,
      session.config.model.clone().as_ref(),
      None,
    ) {
      insta::assert_toml_snapshot!(create_chat_completion_request_message_result);
    } else {
      panic!("construct_chat_completion_request_message failed")
    };
  }
}
