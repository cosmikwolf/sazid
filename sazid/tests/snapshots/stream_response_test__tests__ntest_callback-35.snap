---
source: tests/stream_response_test.rs
expression: "&session.data.rendered_text"
---
"You:\nHello World\nRequest Token Count: 4\nBot: Hello! How can I assist you today? If you have any questions or need\ninformation on a specific topic, feel free to ask.\n"

